### PR TITLE
Add DynSized trait (rebase of #44469)

### PR DESCRIPTION
**Pull request by mikeyhew**
I managed to rebase @plietar's PR #44469, and get all the tests passing.

This unfortunately conflicts with #44917 (by @Zoxc) in a lot of places, so one of us will have to do a rebase after the other gets merged.

r? @pnkfelix since you were reviewing #44469
cc @arielb1

### CODE REVIEW

**Comment by Zoxc** [comment]
Letting `auto trait` imply ?Move and ?DynSized might be a good idea. It would at least avoid all the changes to the tests.

@nikomatsakis Any chance that bounds on `Self` will become sound for auto traits in the future?

**Comment by mikeyhew** [comment]
A couple of interesting points that weren't mentioned in the conversation on #44469:

- `DynSized` is an implicit supertrait of all traits, including `Sized`. This is in contrast to `Sized`, which is only an implicit bound on type parameters and associated types.
-   As a result, this messes up auto traits, which may not have super traits. After this PR, all auto traits must be updated to explicitly remove `DynSized` as  a super trait:

    ```rust
    #![feature(optin_builtin_traits, dynsized)]
    use core::marker::DynSized;
    unsafe auto trait Auto: ?DynSized {}
    ```
- Whenever another implicit supertrait is added, auto traits will break again. This will happen when #44917 lands, which adds an implicit supertrait `Move`.

Because of the potential for future breakage, it may make sense to special-case auto traits, so that they don't have any implicit supertraits. As long as auto traits can't have items, it hopefully should be possible to add or remove implicit supertraits without breaking. Or alternatively, if auto traits could be allowed to have supertraits, we could allow that, as @Zoxc has just mentioned

**Comment by mikeyhew** [comment]
cc @bluss

**Comment by mikeyhew** [comment]
@Zoxc I saw your comment just before I posted mine, just didn't want to rewrite everything. You basically just said the exact same as me, in way fewer words :)

**Comment by mikeyhew** [comment]
ugh... `src/test/rustdoc/foreigntype.rs` is failing, but I don't have that file in my local branch... guess I need to rebase

**Comment by liigo** [comment]
dynamically (at runtime) sized types are `DynSized`, that's ok.
but statically (known at compiletime) sized types are `DynSized`?? that's a bit weird.
Maybe you should rename `DynSized`.

**Comment by mikeyhew** [comment]
@liigo anything that is `Sized` is trivially also `DynSized`. If you know the size of something at compile time, then you know it a runtime

**Comment by mikeyhew** [comment]
@iigo I'll try to explain that a bit better: `DynSized` (in this interpretation) doesn't mean "not statically sized", but rather "has a size that can be calculated at run time".

**Comment by shepmaster** [comment]
Ping from triage @pnkfelix — will you be able to review this shortly?

**Comment by shepmaster** [comment]
Ping from triage, @rust-lang/lang — this PR has gone more than two weeks without a review. Please assign a new reviewer.

File: src/doc/unstable-book/src/language-features/optin-builtin-traits.md [hunk 1]
```diff
@@ -19,12 +19,22 @@ has explictly opted out via a negative impl.
 impl !Type for Trait
 ```
 
+[#46108] added the `DynSized` trait, which is an implicit bound for all traits. Auto traits may not
```
Comment (nikomatsakis) [inline src/doc/unstable-book/src/language-features/optin-builtin-traits.md]: the new move to `auto trait` syntax would make this easier to do automatically, right?

File: src/librustc/ty/mod.rs [hunk 2]
```diff
@@ -1887,6 +1899,74 @@ impl<'a, 'gcx, 'tcx> AdtDef {
...
+        let result = match ty.sty {
+            TyBool | TyChar | TyInt(..) | TyUint(..) | TyFloat(..) |
+            TyRawPtr(..) | TyRef(..) | TyFnDef(..) | TyFnPtr(_) |
+            TyStr | TyDynamic(..) | TySlice(_) | TyArray(..) |
+            TyTuple(..) | TyAdt(..) | TyClosure(..) | TyGenerator(..) |
+            TyNever => {
+                vec![]
+            }
+
+            TyForeign(..) | TyError => {
```
Comment (nikomatsakis) [inline src/librustc/ty/mod.rs]: I believe `TyError` probably wants to be considered `DynSized`

Comment (nikomatsakis) [reply src/librustc/ty/mod.rs]: Well, I guess we treat it the other way for sized-constraint. Probably doesn't matter since we'll just suppress the error we would have reported later on.

File: src/librustc/ty/mod.rs [hunk 3]
```diff
@@ -1887,6 +1899,74 @@ impl<'a, 'gcx, 'tcx> AdtDef {
...
+
+    fn dynsized_constraint_for_ty(&self,
+                                   tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                   ty: Ty<'tcx>)
+                                   -> Vec<Ty<'tcx>> {
+        let result = match ty.sty {
+            TyBool | TyChar | TyInt(..) | TyUint(..) | TyFloat(..) |
+            TyRawPtr(..) | TyRef(..) | TyFnDef(..) | TyFnPtr(_) |
+            TyStr | TyDynamic(..) | TySlice(_) | TyArray(..) |
+            TyTuple(..) | TyAdt(..) | TyClosure(..) | TyGenerator(..) |
```
Comment (nikomatsakis) [inline src/librustc/ty/mod.rs]: I don't understand the treatment of `TyAdt` here. Are we saying that e.g. a `struct` declaration is always `DynSized`?

Comment (nikomatsakis) [reply src/librustc/ty/mod.rs]: (Similarly `TyTuple`, `TyClosure` and `TyGenerator`...)

File: src/doc/unstable-book/src/language-features/optin-builtin-traits.md (hunk 1 above)
Comment (mikeyhew) [reply src/doc/unstable-book/src/language-features/optin-builtin-traits.md]: @nikomatsakis by "do this automatically" do you mean automatically relaxing the default `DynSized` bound for auto traits? I guess so, because with the old way (`impl  Trait for .. {}`) there is no way to tell from the trait definition whether it is an auto trait or not, whereas with `auto trait` it is obvious.

File: src/librustc/ty/mod.rs (hunk 3 above)
Comment (mikeyhew) [reply src/librustc/ty/mod.rs]: @nikomatsakis I think that @plietar decided that `!DynSized` types should not be allowed in struct tails (in the initial implementation), so as a result of that, yes, all structs and tuples are `DynSized`. I can add a FIXME about that though, since it will be changed if `!DynSized` types are allowed in struct tails.

`TyClosure` and `TyGenerator` are currently always `Sized` (see `sized_constraints` above), so they are trivially also `DynSized`.

Comment (nikomatsakis) [reply src/librustc/ty/mod.rs]: OK. A FIXME would be nice, yes.

**Comment by kennytm** [comment]
Could we ensure we *cannot* manually implement `DynSized`?

```rust
extern { type C; }
impl DynSized for C {} //~ ERROR E0322
```

**Comment by mikeyhew** [comment]
@kennytm there is a test for that: https://github.com/rust-lang/rust/pull/46108/files#diff-1d063efd4520e8a566874053ba0440fb

**Comment by kennytm** [comment]
Thanks, missed it, too many files 😄

**Comment by kennytm** [comment]
Triage ping @mikeyhew! 

@nikomatsakis left a few [review comments](https://github.com/rust-lang/rust/pull/46108#pullrequestreview-82277881), primarily requesting a FIXME comment (is there anything else @nikomatsakis?)

Also, please fix the merge conflict.

**Comment by nikomatsakis** [comment]
Well, the bigger question for the review is whether we want this at all. I believe that any ?trait is a very high price to pay and am not convinced that this extension carries its weight (as an alternative, it seems plausible that `sizeof_val` might simply panic). I'm not sure where is best place to have this discussion. On Dec 20, 2017 09:22, kennytm <notifications@github.com> wrote:Triage ping @mikeyhew!
@nikomatsakis left a few review comments, primarily requesting a FIXME comment (is there anything else @nikomatsakis?)
Also, please fix the merge conflict.

**Comment by mikeyhew** [comment]
Thanks for the ping, @kennytm. I'll try to address Niko's comments and rebase this in the next few days

**Comment by mikeyhew** [comment]
@nikomatsakis to start with, we can chat about this on Gitter, and then at some point we may want to create an issue in either this repo or rust-lang/rfcs

**Comment by mikeyhew** [comment]
@nikomatsakis this is a pretty big PR (it adds a lot of code in a lot of files, and complicates the language), and the cited reason for it (disallowing TyForeign in `size_of_val` and other places) is relatively small. I don't think making `size_of_val` panic is a useful long-term solution, but may be useful as a stopgap measure while DSTs are being sorted out.

I'm more interested in the way the `DynSized` trait can be used to make DSTs more usable. As an example, if it had an associated `Meta` type, we could provide methods to split up fat DST pointers into and reassemble from their parts — something you said in https://github.com/rust-lang/rfcs/pull/1524#issuecomment-272020775, except the name of the trait was `Referent`. I have a hacky, experimental crate named `dyn_sized` that provides a similar trait which I use to implement thin pointers in the (also experimental) `thin` crate.

I don't think it is wise to dismiss `?` traits as "not worth the cost"; rather I think we should embrace them, and focus on lowering the cost of adding one to the language. `?` traits are a good way to allow us to add support for types that otherwise couldn't be added without breaking backward compatibility, because they go against assumptions that have been made in the past about all types in Rust. For example, all types are assumed to be movable, but with types like immovable generators, that would not be true. Adding a `Move` trait as a default bound lets us add them to the language, while allowing existing code to work seamlessly as if `!Move` types didn't exist. The same goes for `extern` types and the `DynSized` trait, and there may be other traits that would be useful to add in the future.

I definitely think that DSTs could be improved a lot, and that that will take some serious discussion, design work and at least one RFC (probably an eRFC first). I have some ideas, but am not good enough at written communication to lead that effort, as much as I am interested in it.

**Comment by withoutboats** [comment]
> focus on lowering the cost of adding one to the language

Do you have any ideas of how this could be done?

`?Sized` is one of the most confusing features of the language, that ends up leading people astray with bad error messages and forcing them to put it in places that they don't understand what its doing.  This proposal would now force users to not only deal with `?Sized`, but figure out if it should be `?Sized` or `?DynSized`.

These costs seem inherent to any mechanism that splits every type into one of two boxes like this. Every trait you add increases the number of boxes by a factor of 2. I can't imagine a solution to this, but I'd be very eager to have one.

**Comment by withoutboats** [comment]
And I am also interested in figuring out how we could allow users to define custom dynamically sized types, but that seems orthogonal to having another `?Trait` (you could imagine this `DynSized` trait existing, but being more limited in what it touches).

**Comment by nikomatsakis** [comment]
@mikeyhew 

> I don't think making size_of_val panic is a useful long-term solution

Can you say a bit more about what makes it a bad solution? I can certainly see why it's not the  *ideal* solution, but I'm not yet sure if it's a terrible one. I definitely agree that experience would be good -- i.e., starting with panic semantics and keeping things unstable for a while might help us have concrete examples of why it doesn't work, or it might help us realize that in practice this is Just Fine.

> As an example, if it had an associated Meta type, we could provide methods to split up fat DST pointers into and reassemble from their parts 

I am definitely keen on having reconsidering custom DSTs, but it is not obvious to me why `DynSized` is needed. Perhaps I once knew and have forgotten. =) I'll have to re-read my comment, or maybe you can elaborate on the connection.

> focus on lowering the cost of adding one to the language

Like @withoutboats, I am enthusiastic but mildly skeptical about this idea. =) I agree with you that the benefits of `?traits` are very powerful, and if we can find a way to make them easier to add, that would be great, but I think some of the costs are kind of "inherent". It seems like every `?` trait is going to prompt a re-evaluation of each and every bit of generic context to decide "is this a case where `?Trait` would make sense?"

Even with `?Sized`, I already find it is very common to overlook a `T: ?Sized` bound where one might be legal. The only reliable way to find it is to try and find where they are needed seems to be using unsized types pervasively. And definitely trying to explain `?Sized` can be a challenge.

Actually, it would probably be worth doing a kind of intensive look at `?Sized`, trying to make clear the ergonomic pain points -- maybe some of them can be addressed in other ways (we've had alternative ideas in the past, e.g. I once thought to employ implied bounds, but none of quite panned out), and at least it'd be nice to have concrete examples to talk about.

**Comment by kennytm** [comment]
The main reason for `!DynSized` types is not really about `size_of_val`, but to prevent this from compiling, as stated in #44469:

```rust
extern { type E; }
struct X {
   a: u8,
   b: E,    
   //^ we can't know the alignment of E,
   //  so we can't know offset_of!(X,b) even at runtime
}
```

In particular `Rc<T>` will need `T` to be `DynSized + ?Sized` due to the `RcBox`.

```rust
struct RcBox<T: ?Sized> {
    strong: Cell<usize>,
    weak: Cell<usize>,
    value: T,
}
```

**Comment by arielb1** [comment]
@mikeyhew 

About why this PR is delayed:

1. First, there's a trait-system inference issue that means that without NLL and another fix, this PR "semi-randomly" breaks user code, which means that, as NLL is fairly high priority and this isn't, we should probably delay this PR until NLL stabilizes.
2. Second, as you have seen, adding more "Sized-style" traits has an ergonomics disadvantage which we didn't appreciate when we first seen this PR, and we need to find a compromise about that.

**Comment by mikeyhew** [comment]
There's a lot of comments here (which is a good thing), I'll try my best to address them one by one.

@arielb1 

>First, there's a trait-system inference issue...

You are talking about this issue, right? https://github.com/rust-lang/rust/issues/21974#issuecomment-331886186 I was aware of that, since we talked about it in Gitter, but good to mention it here.

@nikomatsakis regarding making `size_of_val` panic: I just don't think that making a library function so fundamental as that panic, when it can, hopefully without too much pain for the programmer, be made into a compile-time error. It just seems hacky and un-Rusty, and that's why it doesn't seem like the right solution for the long term, but you're right, we don't know yet that it's really that much of a problem, and by making it panic now, we can get feedback from people testing out `extern` types to see what the pain points are. I'm not sure how to deal with the case that @kennytm pointed out without a trait bound though.

@withoutboats and @nikomatsakis regarding Custom DST: I think we should leave that discussion for somewhere on internals for now; for the record though I wasn't talking about Custom DST specifically, which is how we would let people define their own dynamically-sized types, but rather how we could make existing DSTs easier to use – e.g. providing `assemble` and `disassemble` methods. For those to work, you need a `Meta` type for the pointer metadata.

I think that addresses everything except for the cost of adding more `?` traits to the language, which I believe is probably the crux of the issue here. I'll talk about that in my next comment.

**Comment by kennytm** [comment]
@arielb1 I've filed rust-lang/rfcs#2255 and replaced the <kbd>S-blocked</kbd> tag by <kbd>S-waiting-on-team</kbd> for both this and the `Move` PR. 

<kbd>S-blocked</kbd> is intended to be used when a PR is blocked by another PR, or to some extent, an external factor like LLVM bug. In this case, however, we are blocked by a decision, whether to accept having `?DynSized` or not. So I think <kbd>S-waiting-on-team</kbd> is more appropriate. 

In fact I'd rather have these two PRs closed, and someone turn rust-lang/rfcs#2255 into an actual RFC, than having them open and stale forever 🙂.

**Comment by withoutboats** [comment]
> I just don't think that making a library function so fundamental as that panic, when it can, hopefully without too much pain for the programmer, be made into a compile-time error

I think adding a `?trait` is a really major pain for all Rust programmers (unless someone has some insight into how to eliminate the costs). I'd also prefer it be compile time, which is why I'm open to a trans time error - but I think there are good reasons to object to a trans time error as well.

**Comment by eddyb** [comment]
I am curious how many APIs would even be able to support `!DynSized`. Was this discussed somewhere? It's not like `!Move` where you can do a lot with it by-value if you avoid borrowing.

**Comment by kennytm** [comment]
I've roughly glanced the use of `?Sized` in libcore/liballoc/libstd to see  the impact of `DynSized` (I don't care about `?Move` since my focus is custom DST). Actually, most `?Sized` bounds do not need an implicit `DynSized`, since they are mostly only passing pointers around. Those remains are: 

### `T: ?Sized` will remain `T: DynSized + ?Sized`

* `core::intrinsics::{size_of_val, min_align_of_val}`
* `core::mem::{size_of_val, min_align_of_val, align_of_val}`
* `alloc::allocator::Layout::for_value`
* `RefCell` (but `core::cell::{Ref, RefMut}` can be relaxed)
* `Arc` (and `alloc::arc::Weak`)
* `Rc` (and `alloc::rc::Weak`)
* `Box`
* `Mutex`
* `RwLock`

Basically only smart pointers will need `DynSized` bounds.

### Uncertain

Depends on whether we accept `struct Wrapper { value: NonDynSized }`.

* `UnsafeCell`

### Others

* `Index<T>` (and `IndexMut`), the `T` is currently `: ?Sized` due to reasons 😕 

---

Unfortunately, we can't just automatically relax `?Sized` to mean `?DynSized`, because this will break existing code. But this does support what @nikomatsakis [worried](https://github.com/rust-lang/rust/pull/46108#issuecomment-353307253), that *a lot of APIs* can be further relaxed to `?DynSized` since they are just using `?Sized` for forwarding with `&T` or `AsRef`.

**Comment by mikeyhew** [comment]
@kennytm 

>I've roughly glanced the use of `?Sized` in libcore/liballoc/libstd to see the impact of `DynSized` (I don't care about `?Move` since my focus is custom DST)

I just want to be clear, the ergonomic impact of `?DynSized` is caused by the presence of `!DynSized` types, such as `extern` types. Custom DSTs would still implement `DynSized`, so if we had custom DST but no `extern` types, then people would never have to choose whether to write `?DynSized` or just `?Sized`; it would always be `?Sized`.


@kennytm I'm having a hard time following the rest of your comment, do you think you could edit it for clarity? It would be helpful if you explained why each of the types you listed require `T: DynSized`, and also if you added some examples of types that could have `T: ?DynSized`

**Comment by kennytm** [comment]
@mikeyhew 

Everything not listed can be relaxed from `?DynSized`, e.g. `Cow<'a, T>`, `PartialEq<T> for U`, etc.

`Box` needs `DynSized` to get the size for [deallocation](https://doc.rust-lang.org/std/heap/trait.Alloc.html#tymethod.dealloc).

Additionally, as explained in https://github.com/rust-lang/rust/pull/46108#issuecomment-353316536 `Rc` needs `DynSized` due to containing a struct with last field being unsized, thus needing to know the DST's alignment. Same for `Arc`/`Mutex`/`RefCell` etc.

**Comment by mikeyhew** [comment]
Some notes on some the pain points with `?` traits that have been raised:

@withoutboats 
>`?Sized` is one of the most confusing features of the language, that ends up leading people astray with bad error messages and forcing them to put it in places that they don't understand what its doing.

I think a lot of the pain here has to do with the `Sized` trait itself, and the fact that trait objects don't implement it. It's common for people new to Rust to use trait object syntax, and then get an error about `Trait` not implementing `Sized` — [that's what happened to me](https://stackoverflow.com/questions/34898330/slight-tweak-to-custom-fnbox-code-stops-it-from-compiling). `!DynSized` types are a lot rarer, and will hopefully people won't even need to know about the `DynSized` trait until they have relatively advanced knowledge of Rust.

>>focus on lowering the cost of adding one to the language
>
>Do you have any ideas of how this could be done?

OK so when @nikomatsakis was saying "not worth the cost", I thought he meant that it would make the compiler more complicated. I don't have any ideas for reducing the mental headache of having to choose which default bounds to relax, except to suggest that this is a necessary cost of having `!DynSized` or `!Move` types in the language.

**Comment by nikomatsakis** [comment]
I'm torn on this issue. Like, I suspect, all of us, I would obviously prefer to have this error be detected statically. However, I do have grave concerns about the cost of `?Foo` traits, and it's not obvious to me that static detection is worth it (and, it's worth noting, there are compromises, like post-monomorphization *lints* -- which one could set to `deny` if one chose).

Overall, I'd like to elaborate the pros/cons here a bit more rigorously. Thanks to @kennytm for taking the time to analyze the stdlib, at least, and get a feeling for how often people would have to be made aware of this distinction.

That said, @mikeyhew, I am still confused by the references to custom DST. Perhaps I am just being dense, but these feel like somewhat orthogonal questions. I can see that `DynSized` and `Referent` both play a very similar role, in terms of describing the content that a referent can refer to, but it seems like we could have custom DST without an implicit `DynSized` trait. It would mean, I suppose, that `Referent` is implemented for all types (since `&T` is always a legal type), and the compiler knows that. This would include `extern` types -- leaving us with the need to decide what `size_of_val` should do (but that's kind of where we started).

**Comment by mikeyhew** [comment]
@nikomatsakis I guess I see what you are saying... the `Referent` trait could exist regardless of whether the `DynSized` trait exists. Would it look like this?

```rust
trait Referent {
    type Meta: Copy;
}
```

I was thinking that `DynSized` (if it exists) would have the `Meta` associated type, but putting it under `Referent` *would* be more general. Theoretically, there could be types with pointer metadata that aren't `DynSized`<sup>1</sup>.

>However, I do have grave concerns about the cost of ?Foo traits

I really don't think `?DynSized` would be that bad. The only people that have to worry about it are people using `extern` types, and people writing generic code that is used by `extern` types. Sure, it will show up a lot in the standard library, but authors of crates like `serde` can completely ignore the fact that `!DynSized` types exist... and if someone using `extern` types realizes that the crate could support them, they can post an issue, file a pull request, and add support for them in a backward-compatible way.

Can you imagine trying to do that – support `extern` types in your library crate –
 when you're getting trans errors, or worse, panics, possibly in your dependencies? Sure, default trait bounds can be confusing, but I imagine that would be worse. What if a dependency crate adds an "optimization" that uses `size_of_val`, or changes the layout of a struct so that `align_of_val` is required?

The cost of adding `?Move` may be higher, and should be considered separately. I don't think trans errors or panics are a good way of dealing with immovable types either, though.

<hr>

<sup>1</sup> By the way, there really should be two `DynSized` traits: one for types where an `&Self` is needed to determine the size and alignment, and one where only a `<Self as Referent>::Meta` is needed. There are some data structures that require the second trait. Only one of them needs to be a default bound, though.

**Comment by nikomatsakis** [comment]
@mikeyhew 

> Theoretically, there could be types with pointer metadata that aren't DynSized

C-style strings are a common example.

**Comment by nikomatsakis** [comment]
@mikeyhew 

> Can you imagine trying to do that – support extern types in your library crate –
when you're getting trans errors, or worse, panics, possibly in your dependencies? Sure, default trait bounds can be confusing, but I imagine that would be worse. What if a dependency crate adds an "optimization" that uses size_of_val, or changes the layout of a struct so that align_of_val is required?

So, I am not convinced that the cost is small, but I do agree that having the ability to enforce the constraints statically will make maintenance better, and also means that the semver story is much clearer.

**Comment by mikeyhew** [comment]
@nikomatsakis 

>C-style strings are a common example.
>
>EDIT: Well, I guess they can be "sized" via strlen, but it's unclear that size_of_val should be O(n).

Plus, C strings are thin pointers. By "types with pointer metadata" I mean types like `dyn Trait` and `[T]` that are behind fat pointers.

**Comment by nikomatsakis** [comment]
@mikeyhew they would be fat pointers with zero-sized meta data, I think is the idea. This was an example that came up in the Custom DST RFC, IIRC.

**Comment by mikeyhew** [comment]
@nikomatsakis right... then I guess what I meant was types that would have non-zero-sized metadata, but anyway it looks like we're on the same page there.

**Comment by glaebhoerl** [comment]
(If we *don't* end up adding a new `?Trait`, I think it'd be best to not just patch `size_of_val` to either panic or return 0 for `extern type`s, but also at the same time deprecate it and add a new function that properly returns an `Option`?)

**Comment by nikomatsakis** [comment]
@glaebhoerl I also had this plan in mind, yeah. But it doesn't seem to change the fundamental trade-off. I do wonder if people would feel better about it though =)

**Comment by joshtriplett** [comment]
Much worse, not better. It'd be painful to call a method like `size_of_val` on something known at compile time to have a valid size, and get back an `Option`. It should statically give a valid answer (or statically give a compile-time error if called on something it can't work on).

**Comment by nikomatsakis** [comment]
So @withoutboats and I were talking and we had what I think is an interesting insight. We were thinking about when one might want `T` vs `T: ?Sized` vs `T: ?DynSized` and how to relate those to the concepts of ownership / borrowing. We came up with the idea that:

- `T` -- will use this "by value" (no indirection)
- `T: ?Sized` -- may take ownership (or at least partial ownership), but only use through a "smart pointer" (e.g., `Box`, `Arc`)
    - this falls out because, at least with our current allocators, we can't free memory unless we know its size, so you must have `size_of_val`
- `T: ?DynSized` -- no ownership (e.g., only used via `&T` or `*mut T`)
    - if something is borrowed, you don't need to know its size

I'm not sure yet what this means but I find it to be an interesting correspondence. Are there exceptions?

**Comment by mikeyhew** [comment]
- `T: ?Dynsized` can still be owned, just not using `Box`. People will probably implement `Drop` for `T` by calling some C api function, and use a different owning pointer type that just calls `ptr::drop_in_place` when dropped.
- The distinction between `T` and `T: ?Sized` will be clouded a bit if/when we have by-value DST.

**Comment by nikomatsakis** [comment]
@mikeyhew 

> just not using Box

Yes; not using the Rust allocator, basically. Not sure how important that is. (Sometimes I regret designing our allocators so that they supply the size of the value being freed, but I guess it's awfully late to try and reverse course there.)

> The distinction between T and T: ?Sized will be clouded a bit if/when we have by-value DST.

Yes, true.

**Comment by withoutboats** [comment]
> (Sometimes I regret designing our allocators so that they supply the size of the value being freed, but I guess it's awfully late to try and reverse course there.)

It seems like the use cases for size_of_val diminish significantly without this requirement. I wonder if we couldn't revisit this?

**Comment by nikomatsakis** [comment]
OK, I feel bad that this PR is sitting here with no viable means to go forward or backward yet. I remain highly nervous about `?` traits like `?DynSized`. It still feels to me like this is overall a bad direction for Rust to be going in. But I don't want to get into a litany of these concerns (though it makes sense to write those out and work through them).  I would rather focus on the fate of *this PR*.

i.e., do we want to move forward with `?DynSized` even in an experimental capacity? Or should we close the PR?

(It seems clear we have to ensure it is 100% backwards compatible; I've kind of lost track of whether that is the case now.)

@mikeyhew if your main motivation is to explore custom DST, I'd rather just explore that directly, while leaving the questions around `DynSized` slightly unresolved. I continue to feel they are somewhat orthogonal questions.

**Comment by mikeyhew** [comment]
- @nikomatsakis and @withoutboats, and anyone else with concerns about `?DynSized` and other `?`-traits, a good place to discuss them is in https://github.com/rust-lang/rfcs/issues/2255
- Regarding backward compatibility, this PR is in theory backward compatible, but there is still [this issue](https://github.com/rust-lang/rust/issues/21974#issuecomment-331886186), which will cause some regressions. Back in October, @Zoxc listed crates that were affected by that issue in [this comment](https://github.com/rust-lang/rust/pull/44917#issuecomment-335042223)
- about custom DST and related stuff, I would also like to move ahead and experiment with that. What I can do is open a thread on internals and discuss the design that I have in mind, using the `DynSized` trait, and then you can poke and prod and we can discuss alternatives

**Comment by nikomatsakis** [comment]
@mikeyhew 

> a good place to discuss them is in rust-lang/rfcs#2255

agreed, this PR doens't feel right!

> this PR is in theory backward compatible, but there is still this issue, which will cause some regressions

OK, I see, I remember that now. Thanks for the clarification.

> What I can do is open a thread on internals and discuss the design that I have in mind, using the DynSized trait, and then you can poke and prod and we can discuss alternatives

Sounds like a plan.

**Comment by kennytm** [comment]
FYI I've submitted rust-lang/rfcs#2310 as an alternative.

**Comment by nikomatsakis** [comment]
@rfcbot fcp close

We discussed this in the @rust-lang/lang meeting today. As a result of that conversation, I'm going to move to close this PR. The general feeling was that the "jury is still out" on what approach to use for handling `extern` types and the related problems -- as @withoutboats and I have shared here, there is a general concern about the complexity costs of a `?DynSized`-based solution (and `?`-trait-bounds in general), particularly when weighed against the relative obscurity of extern types (it's a lot of language baggage for a corner case). On the other hand, @mikeyhew and others make a strong case for static guarantees. 

Given this uncertainty, I at least am reluctant to land deep changes to the compiler at this time, and in particular not to the trait system (which I think is due for some overhauls). Past experience (e.g., with the `!` type) shows that this often has unexpected consequences, causing regressions and other maintenance burden, which then distracts from higher priority goals. 

Speaking personally: I do intend to leave a comment on rust-lang/rfcs#2255 "real soon now" trying to spell out some of the specific confusion that can arise from `?DynSized`. I am also still intrigued by the possibility of finding an "ownership-based" way of thinking about optional traits [along the lines I outlined before](https://github.com/rust-lang/rust/pull/46108#issuecomment-356755051), though I don't know if that really leads anywhere.

**Comment by shepmaster** [comment]
Ping for ticky boxes @Zoxc, @arielb1, @jseyfried, @michaelwoerister!

**Comment by nikomatsakis** [comment]
I tagged @arielb1 (who should be moved to alumni status, really) as well as @michaelwoerister (who is on vacation afaik) and @jseyfried (who probably also should be moved to alumni status).

**Comment by aturon** [comment]
@nikomatsakis you can modify the FCP list here: https://github.com/anp/rfcbot-rs/blob/master/teams.toml

**Comment by nikomatsakis** [comment]
OK, going to close. Thanks @mikeyhew

**Comment by Ericson2314** [comment]
@nikomatsakis The language team decision against `?DynSize` looked to me like "jurry is still out, so the fact that the code is in flux (chalk, etc) we use as a tie-breaker". That sounds to me more like "not now" than "no never": is there a possibility for landing this on an experimental basis once the type system overhauls are done?

The problem seems to be `?`-traits syntax more than semantics that people want. I say landing this on an experimental basis is best way to find a different syntax / explore downstream problems like custom DSTs. Otherwise too many chained problems are hard to untangle.

**Comment by Ericson2314** [comment]
@mikeyhew See the end of https://github.com/rust-lang/rust/issues/43467 . Looks like `DynSized` is off the table for now, but hopefully it could come back as a non-default bound with `{size,align}_of_bound` being deprecated.
